pkg/coscheduling/coscheduler.go

```diff
@@ -0,0 +1,484 @@
+		}
+		pgs := pgsObj.(*schecache.PodGroupSchedulingStatus)
+		pgs.MatchedPodNodes.Delete(string(p.UID))
+	}
+}
+
+// PostBind is called after a pod is successfully bound. These plugins are
+// informational. A common application of this extension point is for cleaning
+// up. If a plugin needs to clean-up its state after a pod is scheduled and
+// bound, Postbind is the extension point that it should register.
+func (cs *coSchedulingPlugin) PostBind(ctx context.Context, _ *framework.CycleState, p *corev1.Pod, nodeName string) {
+	klog.V(5).Infof("PostBind pod: %v/%v", p.Namespace, p.Name)
+	cs.operation.PostBind(ctx, p, nodeName)
+}
+
+// Less are used to sort pods in the scheduling queue.
+func (cs *coSchedulingPlugin) Less(pi1 *framework.QueuedPodInfo, pi2 *framework.QueuedPodInfo) bool {
+	return cs.operation.Less(pi1, pi2)
+}
+
+// UpdateLocalCache upadte pgstatus cache
+func (cs *coSchedulingPlugin) UpdateBatchCache() {
+	// update pods cache
+	cs.frameworkHandler.IterateOverWaitingPods(func(waitingPod framework.WaitingPod) {
+		klog.V(5).Info("Start IterateOverWaitingPods")
+		pod := waitingPod.GetPod()
+		pgName, satisfied := util.VerifyPodLabelSatisfied(pod)
+		klog.V(5).Infof("Start walking through pod %v/%v uid: %v", pod.Namespace, pod.Name, pod.UID)
+		if satisfied && len(pgName) != 0 {
+			fullName := fmt.Sprintf("%v/%v", pod.Namespace, pgName)
+			pgsObj := cs.operation.PodGroupStatusCache().Get(fullName)
+			if pgsObj == nil {
+				return
+			}
+			pgs := pgsObj.(*schecache.PodGroupSchedulingStatus)
+			klog.V(5).Infof("Wanted cache pod %v/%v", pod.Namespace, pod.Name)
+			oldUID, found := pgs.PodNameUIDs.Get(fmt.Sprintf("%v/%v", pod.Namespace, pod.Name))
+			// pod has been scheduled ever
+			if found {
+				if oldUID.(string) != string(pod.UID) {
+					// delete the expired one
+					pgs.MatchedPodNodes.Delete(oldUID.(string))
+					pgs.PodNameUIDs.Delete(fmt.Sprintf("%v/%v", pod.Namespace, pod.Name))
+					klog.V(3).Infof("Delete old cache data %v", oldUID)
+				}
+			} else {
+				// newly add one
+				klog.V(3).Infof("Add new cache data %v", pod.UID)
+			}
+			klog.V(6).Infof("Current pod group state %+v", pgs.PodNameUIDs.Items())
+			cs.operation.PodGroupStatusCache().Set(util.GetPodGroupFullName(pgs.PodGroup), pgs)
+		}
+	})
+}
+
+// StartBatchSchedule receives the group name of pod to start scheduling
+func (cs *coSchedulingPlugin) StartBatchSchedule(fullName string) {
+	pgs := cs.cache.Get(fullName).(*schecache.PodGroupSchedulingStatus)
+	if pgs.PodGroup.Status.Phase != v1alpha1.PodGroupPreScheduling && pgs.PodGroup.Status.Phase != v1alpha1.
+		PodGroupScheduling {
+		return
+	}
+
+	// record time to avoid abnormal exit when bind
+	if pgs.PodGroup.Status.Scheduled >= pgs.PodGroup.Spec.MinMember {
+		pg, err := cs.client.SchedulingV1alpha1().PodGroups(pgs.PodGroup.Namespace).Get(context.TODO(), pgs.PodGroup.Name,
+			metav1.GetOptions{})
+		if err != nil {
+			klog.Error(err)
+			return
+		}
+		pgCopy := pg.DeepCopy()
```

Comment:
If the obj is from API server, you don't need to do DeepCopy - DeepCopy is only needed from a sharedInformer.

Reply:
Use DeepCopy here just for `Patch`